### PR TITLE
New resources: aws_dx_hosted_transit_virtual_interface and aws_dx_hosted_transit_virtual_interface_accepter

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Part of the work for https://github.com/terraform-providers/terraform-provider-aws/issues/8490.
Related:
* https://github.com/terraform-providers/terraform-provider-aws/pull/8522

Acceptance tests (requires an existing Direct Connect connection in _available_ state):

```console
$ AWS_ALTERNATE_PROFILE=pppppppp DX_CONNECTION_ID=dxcon-aaaaaaaa make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxHostedTransitVirtualInterface_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxHostedTransitVirtualInterface_basic -timeout 120m
=== RUN   TestAccAwsDxHostedTransitVirtualInterface_basic
=== PAUSE TestAccAwsDxHostedTransitVirtualInterface_basic
=== CONT  TestAccAwsDxHostedTransitVirtualInterface_basic
--- PASS: TestAccAwsDxHostedTransitVirtualInterface_basic (646.75s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	646.830s
```

### CODE REVIEW

File: website/docs/r/dx_hosted_transit_virtual_interface.html.markdown [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+* `mtu` - (Optional) The maximum transmission unit (MTU) is the size, in bytes, of the largest permissible packet that can be passed over the connection. The MTU of a virtual transit interface can be either `1500` or `8500` (jumbo frames). Default is `1500`.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The ID of the virtual interface.
+* `arn` - The ARN of the virtual interface.
+* `aws_device` - The Direct Connect endpoint on which the virtual interface terminates.
+* `jumbo_frame_capable` - Indicates whether jumbo frames (9001 MTU) are supported.
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply website/docs/r/dx_hosted_transit_virtual_interface.html.markdown]: Whoops, missed that one; Fixed.

**Comment by ewbankkit** [comment]
Rebased to fix merge conflict.
@n3ph Are there any more review comments to address?

File: website/aws.erb [hunk 2]
```diff
@@ -876,6 +876,12 @@
                                 <li>
                                     <a href="/docs/providers/aws/r/dx_public_virtual_interface.html">aws_dx_public_virtual_interface</a>
                                 </li>
+                                <li>
+                                    <a href="/docs/providers/aws/r/dx_hosted_transit_virtual_interface.html">aws_dx_hosted_transit_virtual_interface</a>
```
Comment (bflad) [inline website/aws.erb]: Nit: The sidebar should be alphabetically sorted

File: website/docs/r/dx_hosted_transit_virtual_interface.html.markdown [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+Provides a Direct Connect hosted transit virtual interface resource.
+This resource represents the allocator's side of the hosted virtual interface.
+A hosted virtual interface is a virtual interface that is owned by another AWS account.
+
+## Example Usage
+
+```hcl
+resource "aws_dx_hosted_transit_virtual_interface" "example" {
+  connection_id = "dxcon-zzzzzzzz"
```
Comment (bflad) [inline website/docs/r/dx_hosted_transit_virtual_interface.html.markdown]: Might be worth showing the possible resource reference here:

```suggestion
  connection_id = "${aws_dx_connection.example.id}"
```

File: website/docs/r/dx_hosted_transit_virtual_interface_accepter.html.markdown [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `dx_gateway_id` - (Required) The ID of the [Direct Connect gateway](dx_gateway.html) to which to connect the virtual interface.
+* `virtual_interface_id` - (Required) The ID of the Direct Connect virtual interface to accept.
+* `tags` - (Optional) A mapping of tags to assign to the resource.
+
+### Removing `aws_dx_hosted_transit_virtual_interface_accepter` from your configuration
```
Comment (bflad) [inline website/docs/r/dx_hosted_transit_virtual_interface_accepter.html.markdown]: This section should instead be a note at the top of the resource page, using the `~> **NOTE:** Message here` syntax

File: aws/resource_aws_dx_hosted_transit_virtual_interface_test.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		for _, v := range resp.VirtualInterfaces {
+			if *v.VirtualInterfaceId == rs.Primary.ID && !(*v.VirtualInterfaceState == directconnect.VirtualInterfaceStateDeleted) {
+				return fmt.Errorf("[DESTROY ERROR] Dx Private VIF (%s) not deleted", rs.Primary.ID)
+			}
+		}
+	}
+	return nil
+}
+
+func testAccCheckAwsDxHostedTransitVirtualInterfaceExists(name string) resource.TestCheckFunc {
```
Comment (bflad) [inline aws/resource_aws_dx_hosted_transit_virtual_interface_test.go]: Same note as #8522 here -- the `testAccCheck...Exists()` should be verifying the infrastructure was properly created via the API -- similar but opposite of `testAccCheckAwsDxHostedTransitVirtualInterfaceDestroy` above. 👍

File: website/aws.erb (hunk 2 above)
Comment (ewbankkit) [reply website/aws.erb]: Will do.

File: website/docs/r/dx_hosted_transit_virtual_interface_accepter.html.markdown (hunk 4 above)
Comment (ewbankkit) [reply website/docs/r/dx_hosted_transit_virtual_interface_accepter.html.markdown]: Will do.

File: aws/resource_aws_dx_hosted_transit_virtual_interface_test.go (hunk 5 above)
Comment (ewbankkit) [reply aws/resource_aws_dx_hosted_transit_virtual_interface_test.go]: Will do.

**Comment by ewbankkit** [comment]
@bflad Review comments addressed.
Similar changes for the acceptance tests as for [`aws_dx_transit_virtual_interface`](https://github.com/terraform-providers/terraform-provider-aws/pull/8522#issuecomment-516552411).

Latest acceptance tests:

```console
$ AWS_ALTERNATE_PROFILE=pppppppp DX_CONNECTION_ID=dxcon-aaaaaaaa make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxHostedTransitVirtualInterface'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxHostedTransitVirtualInterface -timeout 120m
=== RUN   TestAccAwsDxHostedTransitVirtualInterface
=== RUN   TestAccAwsDxHostedTransitVirtualInterface/basic
=== RUN   TestAccAwsDxHostedTransitVirtualInterface/accepterTags
--- PASS: TestAccAwsDxHostedTransitVirtualInterface (1734.33s)
    --- PASS: TestAccAwsDxHostedTransitVirtualInterface/basic (1078.51s)
    --- PASS: TestAccAwsDxHostedTransitVirtualInterface/accepterTags (655.82s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1734.409s
```

**Comment by bflad** [comment]
Thanks, @ewbankkit. Updates look great here too. If another person can verify the acceptance testing also passes on their own Direct Connect connection and show the output here, we'll merge this in. 👍

**Comment by ewbankkit** [comment]
One last change - Setting arn can be moved to the Read method now, removing duplicated code.
This can't be done for `aws_dx_hosted_transit_virtual_interface_accepter` as there's no way to set tags while confirming (accepting) the interface.

**Comment by ewbankkit** [comment]
Using common _Exists_ and _Destroy_ methods from https://github.com/terraform-providers/terraform-provider-aws/pull/9572 in acceptance tests.
Acceptance tests:

```console
$ AWS_ALTERNATE_PROFILE=pppppppp DX_CONNECTION_ID=dxcon-aaaaaaaa make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxHostedTransitVirtualInterface'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxHostedTransitVirtualInterface -timeout 120m
=== RUN   TestAccAwsDxHostedTransitVirtualInterface
=== RUN   TestAccAwsDxHostedTransitVirtualInterface/basic
=== RUN   TestAccAwsDxHostedTransitVirtualInterface/accepterTags
--- PASS: TestAccAwsDxHostedTransitVirtualInterface (1360.23s)
    --- PASS: TestAccAwsDxHostedTransitVirtualInterface/basic (696.04s)
    --- PASS: TestAccAwsDxHostedTransitVirtualInterface/accepterTags (664.18s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1360.308s
```

**Comment by jkodroff** [comment]
@ryno75 You available to test this one?

**Comment by ewbankkit** [comment]
Rebased to fix conflicts and did the Terraform Plugin SDK migration for the new resources.

@ryno75 Will you be able to verify?
Thanks.

**Comment by ewbankkit** [comment]
### TODO

* Any documentation changes for https://github.com/terraform-providers/terraform-provider-aws/pull/10684

**Comment by devonbleak** [comment]
Are we just missing an additional test validation on this?  I have a spare DX connection at the moment and can probably run this tomorrow if so.

**Comment by ewbankkit** [comment]
@devonbleak Yes, an additional test validation is all that is required and would be much appreciated; Thanks.

**Comment by devonbleak** [comment]
Confirming acceptance tests:
```
$ AWS_ALTERNATE_PROFILE=tf-test DX_CONNECTION_ID=dxcon-redacted make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxHostedTransitVirtualInterface'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAwsDxHostedTransitVirtualInterface -timeout 120m
=== RUN   TestAccAwsDxHostedTransitVirtualInterface
=== RUN   TestAccAwsDxHostedTransitVirtualInterface/accepterTags
=== RUN   TestAccAwsDxHostedTransitVirtualInterface/basic
--- PASS: TestAccAwsDxHostedTransitVirtualInterface (1163.86s)
    --- PASS: TestAccAwsDxHostedTransitVirtualInterface/accepterTags (538.47s)
    --- PASS: TestAccAwsDxHostedTransitVirtualInterface/basic (625.38s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       1165.061s
```

**Comment by ewbankkit** [comment]
@devonbleak Many thanks for the independent verification.
@bflad Any chance of getting this merged now? I will address the refactor to use `keyvaluetags` via https://github.com/terraform-providers/terraform-provider-aws/pull/9572.

**Comment by bflad** [comment]
Aside: @devonbleak if any of the other open `service/directconnect` PRs interest your team as well, I believe there are a few others simply pending another testing run: https://github.com/terraform-providers/terraform-provider-aws/pulls?q=is%3Aopen+label%3Aservice%2Fdirectconnect+is%3Apr

**Pull request merged**